harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,519 @@
+    # TODO (anda): find a cleaner way to do this.
+    """
+    aggregated_metrics: Dict[str, Any] = {}
+    for sample in metric_samples:
+        for k, v in sample.items():
+            if isinstance(v, dict):
+                aggregated_metrics[k] = aggregated_metrics.get(k, {})
+                for k1, v1 in v.items():
+                    if isinstance(v1, dict):
+                        raise ValueError("only one level of nested is supported")
+                    aggregated_metrics[k][k1] = aggregated_metrics[k].get(k1, 0) + v1
+            else:
+                aggregated_metrics[k] = aggregated_metrics.get(k, 0) + v
+
+    for k, v in aggregated_metrics.items():
+        if isinstance(v, dict):
+            for k1, v1 in v.items():
+                aggregated_metrics[k][k1] = v1 / len(metric_samples)
+        else:
+            aggregated_metrics[k] = v / len(metric_samples)
+
+    return aggregated_metrics
+
+
+class _MetricGroupCollector(metaclass=abc.ABCMeta):
+    """Abstract class that samples and collects groups of system metrics.
+
+    This class is subclassed by metric groups that implement their respective metrics collection
+    logic.
+    """
+
+    def __init__(self) -> None:
+        self.metric_samples: List[Dict[str, Any]] = []
+
+    @property
+    @abc.abstractmethod
+    def group(self) -> str:
+        pass
+
+    def aggregate(self) -> Dict[str, Any]:
+        """Merge the list of `self.metric_samples` into a single dictionary with aggregate values.
+
+        This method should return a single dictionary where the values represent meaningful
+        aggregation for this metric group.
+
+        By default, this averages all the values across `self.metric_samples` by keys. This should
+        be the aggregation method for most if not all metrics, but individual metric group
+        collectors should override this method should they need an alternate aggregation method.
+        """
+        return _average_metric_samples_depth_one(self.metric_samples)
+
+    def reset(self) -> None:
+        self.metric_samples = []
+
+    @abc.abstractmethod
+    def sample_metrics(self) -> None:
+        """Sample all metrics for this group.
+
+        Records metrics as a dictionary mapping each metric name to its metric value and appends
+        this value to the samples on this class (``self.metric_samples``).
+
+        Certain metrics may be associated with additional labels (i.e. GPU UUIDs) in which case
+        the recorded dictionary should be nested with the label as keys. For example:
+
+        .. code:: python
+            {
+                "GPU-UUID-1": {
+                    "gpu_util": 0.12,
+                    "gpu_free_memory": 123.45,
+                },
+                "GPU-UUID-2": {
+                    "gpu_util": 0.23,
+                    "gpu_free_memory": 234.56,
+                }
+            }
+        """
+        pass
+
+
+class _Network(_MetricGroupCollector):
+    group = "network"
```

Comment:
I'm a little surprised this works. I'd have expected that it had to look like
```python
@property
def group:
  return "network"
```

My typechecker doesn't like it, either.

Reply:
I'm a little unsettled by this. I don't think the implemented code works, either, and that it hasn't resulted in failed tests makes me wonder if something is wrong with tests.

currently implemented:
```python
    def group(self) -> str:
        return "network"
```

Reply:
fixed